### PR TITLE
Unify view of the world for the fake clients

### PR DESCRIPTION
**Pull request by twz123**
## Description

In order for the typed clients to share the same view of the world as the dynamic client, they all must share the same (wrapped) object tracker. Create typed clients using the dynamic client's object tracker and transform between typed and unstructured objects.

Use the auto-generated runtime schemes in the fake discovery client. The client is usually auto-detecting the list types from the scheme. By not passing a new, empty scheme, but the default scheme, there's no longer the need to manually manage the item to list type map. Support API resource discovery by taking all the GVKs out of the scheme and convert those to some API resource lists to be added to the discovery client. The conversion is just a best guess, but will suffice for testing needs.

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update

## How Has This Been Tested?

- [x] Manual test
- [ ] Auto test added

## Checklist:

- [x] My code follows the style [guidelines](https://github.com/k0sproject/k0s/blob/main/docs/contributors/overview.md) of this project 
- [x] My commit messages are [signed-off](https://github.com/k0sproject/k0s/blob/main/docs/contributors/github_workflow.md)
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules
- [x] I have checked my code and corrected any misspellings

### CODE REVIEW

File: pkg/component/controller/apiendpointreconciler_test.go [hunk 1]
```diff
@@ -172,7 +172,7 @@ func TestReconcilerWithNeedForUpdate(t *testing.T) {
 	assert.Equal(t, "bar", e.ObjectMeta.Annotations["foo"])
 }
 
-func verifyEndpointAddresses(t *testing.T, expectedAddresses []string, fakeFactory testutil.FakeClientFactory) *corev1.Endpoints {
+func verifyEndpointAddresses(t *testing.T, expectedAddresses []string, fakeFactory *testutil.FakeClientFactory) *corev1.Endpoints {
```
Comment (kke) [inline pkg/component/controller/apiendpointreconciler_test.go:175]: This fakeclientfactory business is somehow hard to follow.

Maybe there should be some kind of `ClientGetter` interface and functions like this one would take `clientFactory ClientGetter` instead of `fakeFactory *testutil.FakeClientFactory`).

Comment (kke) [reply pkg/component/controller/apiendpointreconciler_test.go:175]: It's `clientFactory kubeutil.ClientFactoryInterface` when it is ultimately used.

Comment (kke) [reply pkg/component/controller/apiendpointreconciler_test.go:175]: Accept interfaces, return concrete types :)

File: pkg/component/controller/workerconfig/reconciler_test.go [hunk 2]
```diff
@@ -52,7 +52,7 @@ type kubeletConfig = kubeletv1beta1.KubeletConfiguration
 
 // TODO: simplify it somehow, it is hard to read and to modify, both tests and implementation
 func TestReconciler_Lifecycle(t *testing.T) {
-	createdReconciler := func(t *testing.T) (*Reconciler, testutil.FakeClientFactory) {
+	createdReconciler := func(t *testing.T) (*Reconciler, *testutil.FakeClientFactory) {
```
Comment (kke) [inline pkg/component/controller/workerconfig/reconciler_test.go:55]: The same goes for these.

Maybe they don't need to be paired like this? Perhaps there could be something like:

```go
func createReconciler(t *testing.T, clientFactory ClientGetter) *Reconciler {
	k0sVars, err := config.NewCfgVars(nil, t.TempDir())
	require.NoError(t, err)
	reconciler, err := NewReconciler(
		k0sVars,
		&v1beta1.ClusterSpec{
			API: &v1beta1.APISpec{},
			Network: &v1beta1.Network{
				ClusterDomain: "test.local",
				ServiceCIDR:   "99.99.99.0/24",
			},
		},
		clients,
		&leaderelector.Dummy{Leader: true},
		true,
	)
	require.NoError(t, err)
	reconciler.log = newTestLogger(t)
	return reconciler
}
```

and:

```go
		clients := createFakeClientFactory()
		reconciler := createReconciler(t, clients)
```

Comment (twz123) [reply pkg/component/controller/workerconfig/reconciler_test.go]: Anyways, done in 2c790e52779951665636facdb31b39231a10cc5c. :sweat_smile:

**Review by jnummelin — APPROVED** [review]
